paho/pinger.go

```diff
@@ -81,106 +73,67 @@ func (p *DefaultPinger) Run(conn net.Conn, keepAlive uint16) error {
 		return fmt.Errorf("conn is nil")
 	}
 	p.mu.Lock()
-	if p.timer != nil {
+	if p.running {
 		p.mu.Unlock()
-		return fmt.Errorf("Run() already called")
+		return fmt.Errorf("Run() already in progress")
 	}
-	select {
-	case <-p.done:
+	p.running = true
+	p.mu.Unlock()
+	defer func() {
+		p.mu.Lock()
 		p.mu.Unlock()
-		return fmt.Errorf("Run() called after stop()")
-	default:
+		p.running = false
+	}()
+
+	interval := time.Duration(keepAlive) * time.Second
+	timer := time.NewTimer(0) // Immediately send first pingreq
+	var lastPingSent time.Time
+	for {
+		select {
+		case <-ctx.Done():
+			timer.Stop() // We don't care if the timer has fired
+			return nil
+		case t := <-timer.C:
+			p.mu.Lock()
+			lastPingResponse := p.lastPingResponse
+			pingDue := p.lastPacketSent.Add(interval)
+			p.mu.Unlock()
+
+			if !lastPingSent.IsZero() && lastPingSent.After(lastPingResponse) {
+				p.debug.Printf("DefaultPinger PINGRESP timeout")
+				return fmt.Errorf("PINGRESP timed out")
+			}
+
+			if t.Before(pingDue) {
+				// A Control Packet has been sent since we last checked, meaning the ping can be delayed
+				timer.Reset(t.Sub(pingDue))
```

Comment:
Supposed to be pingDue.Sub(t)?

Reply:
OOps, corrected this

Reply:
Thanks, have moved the Unlock to the end.